# Сборник 2 из 19-ти задач. Соколова Александра  
---
## Из раздела "Я люблю математику"
### Теория вероятностей и мат. статистика

**Задача 1:**
Амеба имеет шанс 25%, 25% и 50% произвести 0, 1 или 2 потомства соответственно. Каждый из потомков амебы также имеет одинаковые вероятности.  
Какова вероятность того, что род амебы вымрет?  

**Решение для задачи 1:** 
Допустим искомая вероятность вымирания рода амебы равна **p**.  
Событие вымирания рода случается если случится одно из трех зависимых событий перечисленных ниже:  
1-ое событие (А) - с вероятностью 25% сразу не будет потомства => P(A) = 1/4 - вероятность события А  
2-ое событие (В)- с вероятностью 25% родиться одна амеба и у нее с вероятностью **p** вымрет род => P(B)=1/4***p**  
3-е событие (С) - с вороятностью 50% родится две амебы и у них вымрут рода одновременно => P(C) = 1/2***p**^2 - поттому что потомки амебы имеют одинаковую с родителем вероятность вымирания рода  
P(события вымирания рода) = P(A) + P(B) + P(C), так как у одной амебы может родиться одновременно максимально два потомка  
P(события вымирания рода) = **p**  

**p** = 1/4 + 1/4***p** + 1/2* **p**^2  
получаем квадратное уравнение: 1/2* **p**^2- 3/4***p** + 1/4  = 0  
2 **p**^2 - 3 **p** + 1 = 0  
D(Determinant) = 3*3 - 4*2*1 = 1  
**p** = (3 ∓ √D)/(2*2) = (3 ∓ 1)/4 = 1 и 0.5  

Ответ **p** = 1 отбрасываем так как это вырожденный случай.  

**Ответ для задачи 1: p= 0.5**   

**Задача 2:**
В любой 15-минутный интервал с вероятностью 20% вы увидите хотя бы одну падающую звезду.  
Какова вероятность того, что вы увидите хотя бы одну падающую звезду в течение часа?  

**Решение для задачи 2:**  
Будем использовать формулу Бернули:  
$P_n(m) = C_n^m p^m q^{(n-m)}$  
$C_n^k = \frac{n!}{k! (n-k)!}$  

Вероятность увидеть звезду в течение 15 минут = 0.2 = p (можем прдположить что это один опыт который проходит в течении 15 минут и вероятность успеха опыта равна p)  
Вероятность не увидеть звезду в течение 15 минут = q = 1 - p = 1 - 0.2 = 0.8  

Проведем n=4 опытов (4*15 минут = 60 минут = 1 час), в каждом из которых может произойти определенное событие («успех») с вероятностью p (или не произойти — «неудача» — с вероятностью q=1-p. Задача — найти вероятность получения ровно m = 0 успехов в этих n=4 опытах.  

$P_4(0) = C_4^0 p^0 q^{(4-0)} = (0.8)⁴ $  

Вероятность того, что вы НЕ увидите НИ одну падающую звезду в течение часа = Вероятность не увидеть ни одной звезды за 15 минут 4 раза подряд = (0.8)⁴ = 0.4096 так как падение звезд в каждый 15-ти минутный промежуток времени независимо друг от друга  

Вероятность того, что вы увидите хотя бы одну падающую звезду в течение часа = 1 - Вероятность того, что вы НЕ увидите НИ одну падающую звезду в течение часа = 1 - 0.4096 = 0,5904  

**Ответ для задачи 2: 0,5904**   

**Дополнительное решение для задачи 2:**  
В сети я увидел еще одно решение этой задачи с помощью формулы Пуассона, но на мой взгляд её использование не собосновано, потому что одно из требований применения формулы достаточно большое n. В некоторых источниках указывается что n должно быть больше 30, а n*p <10. Что в рамках нашей задчи не выполняется  

In [78]:
# решение задачи 2 с помощью формулы Пуассона
import scipy.stats as ss
1-ss.poisson.cdf(0,0.8)

0.5506710358827783

**Задача 3:**
Как вы можете сгенерировать случайное число от 1 до 7, используя только кубик?  

**Решение для задачи 3:**  
Для начала договоримся, что кубик у нас идеальный. При бросании кубика случайная величина — число точек на грани —   
принимает одно из 6-и возможных значений: {1,2,3,4,5,6}. Вероятность выпадения одной точки из шести равна 1/6,  
одинакова для каждой точки, поэтому случайная величина имеет дискретное равномерное распределение со следующими параметрами:  
мат. ожидание = 3.5,  
дисперсия ~= 2.916666,  
стандартное отклонение ~= 1.7078,  
коэффициент асимметрии = 0,  
коэффициент эксцесса ~= -1.2685  

Нам необходимо создать способ транформации вышеописанного распределения в распределение с возможными значениями: {1,2,3,4,5,6,7}.  
Вероятность выпадения одного значения из семи равна 1/7 и одинакова для каждого значения и случайная величина имеет дискретное   
равномерное распределение со следующими параметрами:  
мат. ожидание = 4,  
дисперсия = 4,  
стандартное отклонение = 4,  
коэффициент асимметрии = 0,  
коэффициент эксцесса ~= -1.25  

**Ответ для задачи 3**   
Способ который решает вышеописанную задачу:  
Бросаем кубик 3 раза. Каждый раз интерпретируем выпадаемое на кубике значение, как четное или нечетное, то есть переводим в биты.  
Выпадает четное число это 0 (ноль), выпадает нечетное это - 1 (один).  
Получаем три бита (бит - это один разряд двоичного кода (двоичная цифра)), которые начинаются от '000' и заканчиваются '111'.  
Двоичное число '000' соответствует десятичному числу 0(ноль) ,а двоичное число '111' соответствует десятичному числу 7.  
Получаем пространство элементарных исходов, состоящим из 8-ми (восьми) элементов: {0, 1, 2, 3, 4, 5, 6, 7}.   
Полученные элементарные исходы равновозможны так как исходное распределение дискретное равномерное.  
Но нам нужно не восем, а семь исходов. Поэтому мы договариваемся, что если мы получаем исход со значением 0,   
то мы перебрасываем кубики. Тут важно отметить, что процесс с какой-то вероятностью может не завершиться никогда,   
потому что могут непрерывно выпадать четные числа и мы будем получать 000 и будем вынуждены перебрасывать снова и снова.   
Можно посчитать, что эта вероятность равна 1/2 в степени n, где n кол-во подбрасываний. Что означает что при n больше 20 она будет ничтожно мала (меньше 10**-7)  
Таким образом мы получим пространство исходов {1, 2, 3, 4, 5, 6, 7}  



In [84]:
# давайте проверим этот способ
# сначала смоделируем бросание кубика
from scipy.stats import randint
low, high = 1, 7
mean, var, skew, kurt = randint.stats(low, high, moments='mvsk')  # проверяем параметры дистректного равномерного распределения
                                                                  # которое соответствует бросанию игрального кубика
print('Мат.ожидание: ', mean)
print('Дисперсия: ', var)
print('Стандартное отклонение: ', var**1/2)
print('Коэффициент асимметрии: ', skew)
print('Коэффициент эксцесса: ', kurt)

Мат.ожидание:  3.5
Дисперсия:  2.9166666666666665
Стандартное отклонение:  1.4583333333333333
Коэффициент асимметрии:  0.0
Коэффициент эксцесса:  -1.2685714285714285


In [108]:
# сгенерируем массив из десяти миллионов (10 000 000) бросаний кубика
r = randint.rvs(low, high, size=10000000)
r

array([1, 6, 1, ..., 4, 3, 1])

In [109]:
# проверим уникальные значения, среднее и стандартное отклонение этого массива
import pandas as pd
df_r = pd.DataFrame(data= r)
print('Уникальные значения:', sorted(df_r[0].unique()))
df_r.describe()


Уникальные значения: [1, 2, 3, 4, 5, 6]


0
count  1.000000e+07
mean   3.500421e+00
std    1.708246e+00
min    1.000000e+00
25%    2.000000e+00
50%    4.000000e+00
75%    5.000000e+00
max    6.000000e+00

In [113]:
# посмотрим на вероятности появления каждого значения
df_r[0].value_counts(normalize=True)

6    0.166919
4    0.166746
2    0.166736
1    0.166683
5    0.166511
3    0.166405
Name: 0, dtype: float64

In [114]:
# проверим выполнение классического определения вероятности для полученного распределения
df_r[0].value_counts(normalize=True).sum()

1.0

In [118]:
# теперь создадим функцию которая будет реализовывать наш способ и транформировать одно распределение в другое
def conver_uniform_1_6_to_uniform_1_7 (d_r: np.array):
    new_r = np.zeros(len(d_r))
    count, count_zero = 0, 0
    bite_ = ''
    for i in range(len(d_r)):
        bite_ += str(int(d_r[i]<4))
        if len(bite_) == 3:
            a = int(bite_,2) 
            if a > 0:
                new_r[count] = int(bite_,2)
                count += 1
            else:
                count_zero += 1
                
            bite_ = ''
    return new_r, count, count_zero

In [120]:
new_r, count, count_zero = conver_uniform_1_6_to_uniform_1_7(r)

In [121]:
new_r = new_r[:count]
len(new_r)

2916199

**Замечание**
из 10-ти миллионов подбрасываний мы получили 2916199 значений нового распределения. Получается, что если бы мы не отбрасывали нулевые значения мы должны были получить 10000000/3 = 3333333 значения. А получили 2916199. Значит нули были в 3333333-2916199 = 417134 случаях. Это почти 1/8 (~=417134/3333333) от 3333333 значений, которые мы должны были получить.

In [122]:
count_zero

417134

In [127]:
# теперь проверим параметры полученного распределения и сравним с заданным в условии задачи
df_r_new = pd.DataFrame(data= new_r)
print('Уникальные значения:', sorted(df_r_new[0].unique()))
df_r_new.describe()

Уникальные значения: [1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0]


0
count  2.916199e+06
mean   3.999537e+00
std    1.999918e+00
min    1.000000e+00
25%    2.000000e+00
50%    4.000000e+00
75%    6.000000e+00
max    7.000000e+00